tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -421,3 +424,124 @@ def prepare_ion_recomb_coefficients_nlte_ion(
             total_coll_ion_coefficients,
             total_coll_recomb_coefficients,
         )
+
+    def jacobian_matrix(
+        self,
+        populations,
+        rate_matrix,
+        rate_matrix_index,
+        total_rad_recomb_coefficients,
+        total_coll_ion_coefficients,
+        total_coll_recomb_coefficients,
+    ):
+        """Creates the jacobian matrix used for NLTE ionization solver
+
+        Parameters
+        ----------
+        populations : numpy.array
+            Ion populations, electron density
+        rate_matrix : DataFrame
+            Rate matrix used for NLTE solver.
+        rate_matrix_index : MultiIndex
+            (atomic_number, ion_number, treatment type)
+            If ion is treated in LTE or nebular ionization, 3rd index is "lte_ion",
+            if treated in NLTE ionization, 3rd index is "nlte_ion".
+        total_rad_recomb_coefficients : DataFrame
+            Radiative recombination coefficients grouped by atomic number and ion number.
+        total_coll_ion_coefficients : DataFrame
+            Collisional ionization coefficients(should get multiplied by electron density).
+        total_coll_recomb_coefficients : DataFrame
+            Collisional recombination coefficients(should get multiplied by electron density).
+
+        Returns
+        -------
+        numpy.array
+            Jacobian matrix used for NLTE ionization solver
+        """
+        atomic_numbers = (
+            rate_matrix_index.get_level_values("atomic_number")
+            .unique()
+            .drop("n_e")
+        )
+        index = 0
+        jacobian_matrix = rate_matrix.copy().values
+        jacobian_matrix[:-1, -1] = populations[1:]
+        for atomic_number in atomic_numbers:
+            for i in range(index, index + atomic_number):
+                if rate_matrix_index[i][2] == "nlte_ion":
+                    jacobian_matrix[i, -1] = self.deriv_matrix_block(
+                        atomic_number,
+                        total_rad_recomb_coefficients.loc[(atomic_number,)],
+                        total_coll_ion_coefficients.loc[(atomic_number,)],
+                        total_coll_recomb_coefficients.loc[(atomic_number,)],
+                        populations[index : index + atomic_number + 1],
+                        populations[-1],
+                    )[i - index]
+            index += atomic_number + 1
+            jacobian_matrix[index - 1, -1] = 0
+        return jacobian_matrix
```

Comment:
How often is this function called? Can it be compiled with numba and/or use vectorization?

Reply:
It'll be used in the solver, so I assume a lot? If in that case it's better to vectorize it, I can look into it.

Reply:
Yes, anything with lots of calls should be optimized if it's *easy*, but if it takes you a lot of time to work on, add a TODO comment instead.

Reply:
I think that has not be added - but it should just be an issue. making sure we have the labels right.

Reply:
I added it as an issue #2164